plan/stats.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/pingcap/tidb/expression"
+)
+
+type statsProfile struct {
+	count       float64
+	cardinality []float64
+}
+
+// Sort will call this.
+func (p *baseLogicalPlan) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	p.profile = p.basePlan.children[0].(LogicalPlan).statsProfile()
+	return p.profile
+}
+
+func (p *DataSource) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	p.profile = &statsProfile{
+		count:       float64(p.statisticTable.Count),
+		cardinality: make([]float64, len(p.statisticTable.Columns)),
+	}
+	for i, col := range p.Columns {
+		p.profile.cardinality[i] = float64(p.statisticTable.Columns[col.ID].NDV)
+	}
+	return p.profile
+}
+
+func (p *Selection) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	// TODO: estimate it by histogram.
+	p.profile = &statsProfile{
+		count:       childProfile.count * selectionFactor,
+		cardinality: make([]float64, len(childProfile.cardinality)),
+	}
+	for i := range p.profile.cardinality {
+		p.profile.cardinality[i] = childProfile.cardinality[i] * selectionFactor
+	}
+	return p.profile
+}
+
+func (p *Union) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	p.profile = &statsProfile{
+		cardinality: make([]float64, p.schema.Len()),
+	}
+	for _, child := range p.children {
+		childProfile := child.(LogicalPlan).statsProfile()
+		p.profile.count += childProfile.count
+		for i := range p.profile.cardinality {
+			p.profile.cardinality[i] += childProfile.cardinality[i]
+		}
+	}
+	return p.profile
+}
+
+func (p *Limit) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	p.profile = &statsProfile{
+		count:       float64(p.Count),
+		cardinality: make([]float64, len(childProfile.cardinality)),
+	}
+	if p.profile.count > childProfile.count {
+		p.profile.count = childProfile.count
+	}
+	for i := range p.profile.cardinality {
+		p.profile.cardinality[i] = childProfile.cardinality[i]
+		if p.profile.cardinality[i] > p.profile.count {
+			p.profile.cardinality[i] = p.profile.count
+		}
+	}
+	return p.profile
+}
+
+func (p *TopN) statsProfile() *statsProfile {
+	if p.profile != nil {
+		return p.profile
+	}
+	childProfile := p.children[0].(LogicalPlan).statsProfile()
+	p.profile = &statsProfile{
+		count:       float64(p.Count),
+		cardinality: make([]float64, len(childProfile.cardinality)),
+	}
+	if p.profile.count > childProfile.count {
+		p.profile.count = childProfile.count
+	}
+	for i := range p.profile.cardinality {
+		p.profile.cardinality[i] = childProfile.cardinality[i]
+		if p.profile.cardinality[i] > p.profile.count {
+			p.profile.cardinality[i] = p.profile.count
+		}
+	}
+	return p.profile
+}
+
+func getCardinality(cols []*expression.Column, schema *expression.Schema, profile *statsProfile) float64 {
```

Comment:
Add comments about the algorithm.
It seems like just get the max cardinality of all the columns, this doesn't apply to group by columns.